include/seqan3/io/sequence/sequence_file_in.hpp

```diff
@@ -1,175 +1,934 @@
+#include <seqan3/alphabet/quality/illumina18.hpp>
+#include <seqan3/alphabet/quality/quality_composition.hpp>
+#include <seqan3/core/metafunction/basic.hpp>
+#include <seqan3/io/concept.hpp>
+#include <seqan3/io/exception.hpp>
+#include <seqan3/io/filesystem.hpp>
+#include <seqan3/io/record.hpp>
+#include <seqan3/io/detail/in_file_iterator.hpp>
+#include <seqan3/io/detail/record.hpp>
+#include <seqan3/io/sequence/sequence_file_format_concept.hpp>
+#include <seqan3/io/sequence/sequence_file_format_fasta.hpp>
+#include <seqan3/range/container/concatenated_sequences.hpp>
 
 namespace seqan3
 {
 
-// ==================================================================
-// sequence_file_in_traits
-// ==================================================================
+// ----------------------------------------------------------------------------
+// sequence_file_in_traits_concept
+// ----------------------------------------------------------------------------
 
+/*!\interface seqan3::sequence_file_in_traits_concept <>
+ * \brief The requirements a traits_type for seqan3::sequence_file_in must meet.
+ * \ingroup sequence
+ */
+/*!\name Requirements for seqan3::sequence_file_in_traits_concept
+ * \brief You can expect these **member types** of all types that satisfy seqan3::sequence_file_in_traits_concept.
+ * \memberof seqan3::sequence_file_in_traits_concept
+ *
+ * \details
+ *
+ * Note that the alphabet type of the seqan3::field::SEQ_QUAL cannot be specified directly, it is always
+ * seqan3::qualified<sequence_alphabet, quality_alphabet> and the container type templates for
+ * the field are those of seqan3::field::SEQ.
+ *
+ * \{
+ */
+/*!\typedef using sequence_alphabet
+ * \memberof seqan3::sequence_file_in_traits_concept
+ * \brief Alphabet of the characters for the seqan3::field::SEQ; must satisfy seqan3::alphabet_concept.
+ */
+/*!\typedef using sequence_legal_alphabet
+ * \memberof seqan3::sequence_file_in_traits_concept
+ * \brief Intermediate alphabet for seqan3::field::SEQ; must satisfy seqan3::alphabet_concept and be convertible to
+ * `sequence_alphabet`.
+ *
+ * \details
+ *
+ * This alphabet can be a superset of `sequence_alphabet` to allow conversion of some characters
+ * without producing an error, e.g. if this is set to seqan3::dna15 and `sequence_alphabet` is set to seqan3::dna5,
+ * 'M' will be an accepted character and automatically converted to 'N', while 'Z' will still be an illegal
+ * character and produce an error.
+ */
+/*!\typedef using sequence_container
+ * \memberof seqan3::sequence_file_in_traits_concept
+ * \brief Type template of the seqan3::field::SEQ, a container template over `sequence_alphabet`;
+ * must satisfy seqan3::sequence_concept.
+ */
+/*!\typedef using sequence_container_container
+ * \memberof seqan3::sequence_file_in_traits_concept
+ * \brief Type template of a column of seqan3::field::SEQ, a container template that can hold multiple
+ * `sequence_container`; must satisfy seqan3::sequence_concept.
```

Comment:
> and the value type must also satisfy seqan3::sequence_concept! 

no, that's why I require a template and not a type. It' doesn't have a value_type.

> Don't we have a container_of_container_concept? :thinking:

we deleted it because you said it was over-engineered :wink: